### PR TITLE
Extension SDK External API demo

### PR DESCRIPTION
**Pull request by bryans99**
**Demonstrates the following extension SDK functionality:**

1. OAUTH2 authorization with Google.
2. Creating, viewing and updating a Google sheet using the Google sheets rest API.

**Not logged in**

image

**OATH2 Login**

image

**Logged in.**

The very first time a new spreadsheet is created and populated from a sample spreadsheet provided by google. Pressing the initialize sheet from source creates another spreadsheet from the sample. The spreadsheet id is stored in the extension context so is reload the next time the extension is accessed.

image

**Update functionality**

Rows can be inserted (below the selected line), edited, moved and deleted.

image

**Edit form** 

image

### CODE REVIEW

**Comment by bryans99** [comment]
Slept on this last night and realized that although perfectly functional its use of hooks is not scalable. Going to do a little refactoring so hold off reviewing for a bit.

**Review by jkaster — CHANGES_REQUESTED** [review]
wrong name?

**Comment by bryans99** [comment]
This is ready for review again.

No changes to UI. Refactor of OAUTH and SHEETS hooks into providers with contexts which provide better state management.

Additionally, demo specific logic has been removed from the providers to make it easier to use in other extensions.

File: react/javascript/demo-external-api/README.md
Comment (izzymiller) [inline react/javascript/demo-external-api/README.md:62]: Something that has always slightly bugged me about the format of these READMEs where there are .env files to modify— You must restart your dev server to pick up .env changes (i think).

It's no big deal, but we should either mention that you have to restart the dev server or shift the .env setup to before you `yarn develop`

File: react/javascript/demo-external-api/README.md
Comment (izzymiller) [inline react/javascript/demo-external-api/README.md:64]: I don't think this is the most helpful link / maybe not what you were going for. Maybe this? https://developers.google.com/workspace/guides/create-credentials#web Bit sparse but actually has the credential setup instructions...

Comment (izzymiller) [inline react/javascript/demo-external-api/README.md:64]: We should be nice and tell them up front to add https://your.looker.com/extensions/oauth2_redirect to the authorized redirect URI list when creating the client ID.

File: react/javascript/demo-external-api/manifest.lkml
Comment (izzymiller) [inline react/javascript/demo-external-api/manifest.lkml:18]: Extra entitlement? nbd just got me wondering "where in this is the clipboard leveraged?"

Comment (izzymiller) [reply react/javascript/demo-external-api/manifest.lkml:18]: Also, why bother leaving in all of the `no` entitlements? Is there a difference between an omission and an explicit no?

File: react/javascript/demo-external-api/package.json
Comment (izzymiller) [inline react/javascript/demo-external-api/package.json:16]: Ha an artifact of create looker extension. maybe a feature to make better...
```suggestion
  "author": "Looker",
```

File: react/javascript/demo-external-api/package.json
Comment (izzymiller) [inline react/javascript/demo-external-api/package.json:17]: Another create looker extension artifact... this is MIT.
```suggestion
  "license": "MIT",
```

File: react/javascript/demo-external-api/src/components/Sheets.jsx
Comment (izzymiller) [inline react/javascript/demo-external-api/src/components/Sheets.jsx:45]: Are we sure that anyone can access this, or will non @google.com folks get an error?

Comment (izzymiller) [reply react/javascript/demo-external-api/src/components/Sheets.jsx:45]: Perhaps just something to call out to change in the readme, if we can't share globally... Will also need to change the range declaration

**Review by izzymiller — CHANGES_REQUESTED** [review]
This looks really good! Some README questions and I'd want to make sure that spreadsheet source is publicly accessible but the code overall LGTM!

Super handy to have this example to point at.

File: react/javascript/demo-external-api/manifest.lkml
Comment (bryans99) [reply react/javascript/demo-external-api/manifest.lkml]: Fixed. I like to show all available entitlements so that they can see what is available.

File: react/javascript/demo-external-api/src/components/OauthProvider.jsx
Comment (whscullin) [inline react/javascript/demo-external-api/src/components/OauthProvider.jsx:79]: We should probably enable `ignoreDestructuring: true` in projects that use the API to avoid having to one-off these sorts of things.

File: react/javascript/demo-external-api/README.md
Comment (izzymiller) [inline react/javascript/demo-external-api/README.md:3]:
```suggestion
This example demonstrates usage of external API in an extension using javascript. It uses OAUTH2 to get an access token from Google. It then uses that token to create and update a spreadsheet using the Google sheets APIs. The key things to observe are:
```

File: react/javascript/demo-external-api/src/components/Sheets.jsx
Comment (whscullin) [reply react/javascript/demo-external-api/src/components/Sheets.jsx:45]: Either way maybe something for the default `.env` config.

File: react/javascript/demo-external-api/src/components/SheetsForm.jsx
Comment (whscullin) [inline react/javascript/demo-external-api/src/components/SheetsForm.jsx:223]: Could we add a json loader to sort of separate out concerns?

File: react/javascript/demo-external-api/src/components/SheetsProvider.jsx
Comment (whscullin) [inline react/javascript/demo-external-api/src/components/SheetsProvider.jsx:148]: Could `invokeSheetsApi()` do a little more lifting so that the salient parts of the actions is clearer?

**Review by izzymiller — APPROVED** [review]
README updates look great :)

File: react/javascript/demo-external-api/README.md
Comment (jkaster) [inline react/javascript/demo-external-api/README.md:17]:
```suggestion
6. Click `Library` and search for `Sheets`. This brings up a link to the `Google Sheets API`. `Select` it.
```

File: react/javascript/demo-external-api/README.md
Comment (jkaster) [inline react/javascript/demo-external-api/README.md:19]: This is what I see after `Enable`
image

So I think I need to create the credentials first?

Comment (jkaster) [reply react/javascript/demo-external-api/README.md:19]: step 1
image

Comment (jkaster) [reply react/javascript/demo-external-api/README.md:19]: After `Done` I get 
image

Comment (jkaster) [reply react/javascript/demo-external-api/README.md:19]: I created some entries there, then I was able to proceed (for the most part the same way) with step 8.

File: react/javascript/demo-external-api/README.md
Comment (jkaster) [inline react/javascript/demo-external-api/README.md:22]: I clicked `Save and Continue` twice.

File: react/javascript/demo-external-api/README.md
Comment (jkaster) [inline react/javascript/demo-external-api/README.md:24]:
```suggestion
13. Click the `Create Credentials` button and select `OAuth client ID`.
```

File: react/javascript/demo-external-api/README.md
Comment (jkaster) [inline react/javascript/demo-external-api/README.md:28]:
```suggestion
17. Add the base url of the Looker instance that will host your extension appended with `/extensions/oauth2_redirect` as an `Authorized redirect URI`. Example `https://myinstance.looker.com/extensions/oauth2_redirect`.
```

**Review by jkaster — CHANGES_REQUESTED** [review]
Up to "Getting started for development"

File: react/javascript/demo-external-api/README.md
Comment (bryans99) [reply react/javascript/demo-external-api/README.md:19]: You should not have to do anything with Service accounts

File: react/javascript/demo-external-api/src/components/Sheets.jsx
Comment (bryans99) [reply react/javascript/demo-external-api/src/components/Sheets.jsx:45]: I've added a comment to the readme for now.

File: react/javascript/demo-external-api/src/components/OauthProvider.jsx
Comment (bryans99) [reply react/javascript/demo-external-api/src/components/OauthProvider.jsx:79]: I've had to use
`camelcase: ['error', { ignoreDestructuring: true, properties: 'never' }],`

File: react/javascript/demo-external-api/src/components/SheetsProvider.jsx
Comment (bryans99) [reply react/javascript/demo-external-api/src/components/SheetsProvider.jsx]: Yes it could and it now does. I could improve things more but I dont want to re-invent big G's api

File: react/javascript/demo-external-api/src/components/SheetsForm.jsx
Comment (bryans99) [reply react/javascript/demo-external-api/src/components/SheetsForm.jsx]: Done

File: react/javascript/demo-external-api/README.md
Comment (bryans99) [reply react/javascript/demo-external-api/README.md]: Done

File: react/javascript/demo-external-api/README.md
Comment (bryans99) [reply react/javascript/demo-external-api/README.md]: Went through the whole process again. I think I have the definitive version now. I added some comments that hopefully give a hint as to how to fix things if things go wrong

File: react/javascript/demo-external-api/README.md
Comment (jkaster) [inline react/javascript/demo-external-api/README.md:8]: Good discovery. We should use this magic id for the hackathon app startup too.

File: react/javascript/demo-external-api/README.md
Comment (jkaster) [inline react/javascript/demo-external-api/README.md:28]: like that url name change

File: react/javascript/demo-external-api/README.md
Comment (jkaster) [inline react/javascript/demo-external-api/README.md:32]:
```suggestion
19. Enter `Sheets` in the search field. This displays a link to the `Google Sheets API`. Click the link.
```

**Review by jkaster — CHANGES_REQUESTED** [review]
One typo to fix that I'll merge after this review is posted

**Review by jkaster — APPROVED** [review]
LGTM. Great step write up. I was able to get this demo working yesterday with a personal Google account.

File: react/javascript/demo-external-api/README.md
Comment (bryans99) [reply react/javascript/demo-external-api/README.md:28]: Let me double check that! Thats what I get for doing global change!

**Review by jkaster — APPROVED** [review]

**Pull request merged**
